twpca/crossval.py

```diff
@@ -0,0 +1,212 @@
+
+    Args:
+        data: array-like, trials x time x neurons dataset
+        n_components: sequence of ints, number of components in each model
+        warp_scales: sequence of floats, scale of warp regularization in each model
+        time_scales: sequence of floats, scale of time factor regularization in each model
+
+    Keywork Args:
+        warp_reg: function, takes scalar and outputs a regularization term (in tensorflow) for
+                    the warps. By default, warp_reg = twpca.regularizers.curvature(s, power=1).
+        time_reg: function, takes scalar and outputs a regularization term (in tensorflow) for
+                    the time factors. By default,
+                    time_reg = twpca.regularizers.curvature(s, power=2, axis=0).
+        crossval_method: str, specifies cross validation. One of {'kfold' (default), 'leavout'}.
+        K: int, cross validation parameter. For example, K = 5 specifies 5-fold cross validation
+                    when crossval_method = 'kfold', and K = 1 specifies leave-1-out validation
+                    when crossval_method = 'leaveout'
+        max_fits: int, number of fits per model. For example max_fits = 1 means that only a single
+                  training and test set are evaluated (default: np.inf).
+        fit_kw: dict, keyword arguments passed to model.fit
+        **model_kw: additional keywords are passed to twpca.TWPCA(...)
+    """
+
+    # defaults for warp and time regularization
+    if warp_reg is None:
+        warp_reg = lambda s: curvature(scale=s, power=1)
+    if time_reg is None:
+        time_reg = lambda s: curvature(scale=s, power=2, axis=0)
+
+    # initialize results dict
+    results = {
+        'n_components': [],
+        'warp_scale': [],
+        'time_scale': [],
+        'crossval_data': [],
+        'mean_test': [],
+        'mean_train': [],
+        'mean_dim_change': []
+    }
+
+    # run cross-validation for all specified hyperparameters
+    for nc, ws, ts in zip(tqdm(n_components), warp_scales, time_scales):
+
+        model = TWPCA(nc, warp_regularizer=warp_reg(ws), time_regularizer=time_reg(ts), **model_kw)
+        _result = cross_validate(model, data, crossval_method, K, max_fits=max_fits, **fit_kw)
+
+        results['n_components'].append(nc)
+        results['warp_scale'].append(ws)
+        results['time_scale'].append(ts)
+        results['crossval_data'].append(_result)
+        results['mean_test'].append(np.mean([r['test_error'] for r in _result]))
+        results['mean_train'].append(np.mean([r['train_error'] for r in _result]))
+        results['mean_dim_change'].append(np.mean([r['warped_rank']-r['unwarped_rank'] for r in _result]))
+
+    return results
+
+def error_per_component(data, component_range,
+                        fit_kw=dict(lr=(1e-1, 1e-2), niter=(250, 500), progressbar=False),
+                        **model_args):
+    """Compares vanilla PCA to twPCA searching over number of components
```

Comment:
This isn't exported. I'm still playing with the functionality I want - will remove for now.